### PR TITLE
Packaging improvements

### PR DESCRIPTION
**Pull request by robioson**
# Background
#### Anything else the reviewers should know about the changes here
Please test on a mac by running `make uninstall && make install` and testing the stuff you usually use the utils for!

# Changes
### Code changes
This PR incorporates several improvements to the packaging of the utility:
- Support for other Unixes (i.e. Linux)
    - The loader script now installs into `/usr/bin`. Previously it used `/usr/local/bin`, which is macOS specific. The equivalent on most Linux distributions is `~/.local/bin` but this is often not on `PATH`. I don't believe there is an easy way to get the local bin folder in a cross-platform way anyway.
- Installation can now be done with pip
    - The `biomage` package is included in the distribution
    - So is `config.yaml`
    - The entry point is configured correctly. The utility still installs into a venv as before, but making it system-wide is now as simple as symlinking `venv/bin/biomage` to `/usr/bin/biomage`
- Added new make targets: `uninstall`, `develop`
- Changed `fmt`  and `check` to use `venv` binaries so these tools don't have to be installed system-wide
- Changed one instance of Python 3.8 syntax in `stage.py` to make it compatible with 3.7 (see my reasoning in the commit). This was the only logic change.
- Finally, it fixes the CI workflow.

# Definition of DONE
Your changes will be ready for merging after each of the steps below have been completed:

### Testing
- [ ] New commands have been added to Makefile/test
- [ ] Unit tests written
- [x] Run make test
- [x] Test any modified command

To set up easy local testing with inframock, follow the instructions here: https://github.com/biomage-ltd/inframock
To deploy to the staging environment, follow the instructions here: https://github.com/biomage-ltd/biomage-utils

### Documentation updates
Is all relevant documentation updated to reflect the proposed changes in this PR?

- [x] Relevant Github READMEs updated
- [ ] Relevant wiki pages created/updated

### Approvers
- [ ] Approved by a member of the core engineering team

### Optional
- [ ] Photo of a cute animal attached to this PR

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -7,36 +7,54 @@
 # Variables
 #--------------------------------------------------
 PYTHON_FILES?=$$(find biomage -name '*.py')
-INSTALL_PATH=/usr/local/bin
-MKFILE_PATH := $(abspath $(lastword $(MAKEFILE_LIST)))
-MKFILE_DIR := $(dir $(MKFILE_PATH))
+ENTRY_POINT=/usr/bin/biomage
```
Comment (kafkasl) [inline Makefile:10]: you cannot install into `/usr/bin` in mac because it's protected by SIP. Either detect if the OS is linux / mac and install into different directories or just add a note in the `Readme.md` telling linux users to add it to the PATH.

File: Makefile [hunk 2]
```diff
@@ -7,36 +7,54 @@
...
-	@black $(PYTHON_FILES)
-	@echo ""
-check: ## Checks code for linting/construct errors
+	@venv/bin/black $(PYTHON_FILES)
+	@echo "    [✓]"
+
+check: develop ## Checks code for linting/construct errors
 	@echo "==> Checking if files are well formatted..."
-	@flake8 $(PYTHON_FILES)
-	@echo "    [✓]\n"
```
Comment (kafkasl) [inline Makefile:34]: why remove the newline?

File: README.md [hunk 3]
```diff
@@ -8,7 +8,7 @@ infrastructure.
 Setup
 -----
 
-After cloning the repository, do the following (might require **sudo** in some distributions, and `pip` > 21.0.1):
```
Comment (kafkasl) [inline README.md:11]: `make install` does not work in MacOS because of the above reasons, but `uninstall` still requires `sudo` so far. Probably won't after moving the install dir, but need to check before remove that).

File: biomage/experiment/copy.py [hunk 4]
```diff
@@ -2,9 +2,9 @@
 
 import click
 
-from utils.config import get_config
-from utils.constants import DEFAULT_SANDBOX, PRODUCTION, STAGING
-from utils.data import copy_experiments_to
+from ..utils.config import get_config
```
Comment (kafkasl) [inline biomage/experiment/copy.py:5]: why is this import change needed?

File: biomage/rotate_ci/rotate_ci.py [hunk 5]
```diff
@@ -1,13 +1,15 @@
-import click
```
Comment (kafkasl) [inline biomage/rotate_ci/rotate_ci.py:1]: the imports are automatically ordered by VS Code using the settings in `biomage-utils.code-workspace`. Either use replicate those settings in your IDE or move to VS Code

File: biomage/stage/stage.py [hunk 6]
```diff
@@ -431,12 +434,7 @@ def paginate_experiments(
                 answers.remove(done_text)
                 page_action = "done"
 
-            chosen_experiments = set(
-                [
-                    *chosen_experiments,
-                    *[experiment.split(" ")[0] for experiment in answers],
-                ]
-            )
+            chosen_experiments = {*chosen_experiments, *[experiment.split(" ")[0] for experiment in answers]}
```
Comment (kafkasl) [inline biomage/stage/stage.py:437]: why did you remove the set?

File: setup.py [hunk 7]
```diff
@@ -1,20 +1,38 @@
-import setuptools
+from setuptools import setup, find_packages
 
-with open("README.md", "r") as fh:
-    long_description = fh.read()
+with open("README.md") as f:
```
Comment (kafkasl) [inline setup.py:3]: in general, I would appreciate not removing `style` data. imo having the file direction specified is good or at least it does no harm.

**Review by kafkasl — CHANGES_REQUESTED** [review]
The only blockers for the PR are:
* The install directory `/usr/bin/` which makes this unusable in MacOS. 
* The different import ordering (if you have any other different autoformatting-related setting it might be a good moment to change that too)

The rest are just comments that you can choose to address or not.

File: biomage/stage/stage.py (hunk 6 above)
Comment (robioson) [reply biomage/stage/stage.py:437]: It's still a set, I just used a set literal as it's faster and shorter.

File: biomage/experiment/copy.py (hunk 4 above)
Comment (robioson) [reply biomage/experiment/copy.py:5]: Relative imports are far more reliable and should be used wherever possible. When I installed the package with pip, absolute imports didn't work.

File: Makefile (hunk 2 above)
Comment (robioson) [reply Makefile:34]: Newlines are printed literally on Linux, so I thought this was an error. I can fix it by using `-e`, if that works on mac.

File: Makefile (hunk 1 above)
Comment (robioson) [reply Makefile:10]: Will do the former.

File: biomage/rotate_ci/rotate_ci.py (hunk 5 above)
Comment (robioson) [reply biomage/rotate_ci/rotate_ci.py:1]: I can see the line in the workspace but are you sure this is doing the right thing? The ordering seems to be random (not even alphabetic). I manually reordered them into the normal scheme (alphabetic as commanded in PEP8, followed by splitting into stdlib/third party/own groups as recommended by most style guides ([like this](https://github.com/reddit-archive/reddit/wiki/PythonImportGuidelines#ordering-of-import-statements)).

File: Makefile (hunk 1 above)
Comment (kafkasl) [reply Makefile:10]: there's an example in some other makefile iirc so it's consistent, if you find a better way we can change the other

File: biomage/experiment/copy.py (hunk 4 above)
Comment (kafkasl) [reply biomage/experiment/copy.py:5]: awesome, thanks for the explanation

File: biomage/stage/stage.py (hunk 6 above)
Comment (kafkasl) [reply biomage/stage/stage.py:437]: oh right, hadn't seen that notation 👌

File: biomage/rotate_ci/rotate_ci.py (hunk 5 above)
Comment (kafkasl) [reply biomage/rotate_ci/rotate_ci.py:1]: mmm fair enough we'll need to have a look. but let's find some tool, doing it manually it's a mess as they will keep changing depending on commiter

File: Makefile (hunk 1 above)
Comment (robioson) [reply Makefile]: This should be fixed now!

File: biomage/rotate_ci/rotate_ci.py (hunk 5 above)
Comment (robioson) [reply biomage/rotate_ci/rotate_ci.py:1]: Yup, that sounds good to me.

File: Makefile (hunk 2 above)
Comment (robioson) [reply Makefile:34]: Fixed just by adding a blank `echo` under each tick.

File: biomage/rotate_ci/rotate_ci.py (hunk 5 above)
Comment (robioson) [reply biomage/rotate_ci/rotate_ci.py:1]: I added `isort` to the development requirements, and `make fmt` will run it, automatically sorting all imports.

**Review by kafkasl — APPROVED** [review]
I added some sort settings to make it work exactly like the VS code auto-organize (which I have readded).

**Review by kafkasl — APPROVED** [review]

File: biomage/rotate_ci/rotate_ci.py (hunk 5 above)
Comment (marcellp) [reply biomage/rotate_ci/rotate_ci.py:1]: To be clear, from the changes to the workspace file I assume this is also automatically done when you save the files in VS Code manually, correct?

File: README.md (hunk 3 above)
Comment (marcellp) [reply README.md:11]: Has this been fixed since @robioson?

Comment (robioson) [reply README.md:11]: In that you never have to run `sudo make`, yes

File: biomage/rotate_ci/rotate_ci.py (hunk 5 above)
Comment (robioson) [reply biomage/rotate_ci/rotate_ci.py:1]: Pol added this, but yes.
